### PR TITLE
feat: UI optimizations

### LINKED ISSUE #2711: UIShape polling overhead
`UIShape` component has a silly polling algorithm called `LayoutRefreshWatcher` that doesn't really scale when there are more than 1000 UIShapes.

- [ ] Change the `LayoutRefreshWatcher` to avoid Polling by treating the isLayoutDirty as an event rather than a boolean.
- [ ] Go to a scene with +1000 UIShapes and test that the performance drop is gone (try ICE Poker -100,126 )

### PR DESCRIPTION
**Pull request by Kinerius**
## What does this PR change?

- Removes the refresh polling that degrades the performance when there are a lot of UIShapes, in its place now there's a unique coroutine that checks if there's a dirty UIShape on its queue.
- Removed a constant Canvas Rebuild caused by a TextShape being updated every frame with no changes on its data.

Fixes #2711

### Old ( ICE Poker - The Stronghold ) [No Avatars]
```
 * PERFORMANCE SCORE (0-100) -> 64
 * lowest frame time -> 38.9ms
 * average frame time -> 45.8ms
 * highest frame time -> 100.0ms
 * 1 percentile frame time -> 39.3ms
 * 50 percentile frame time -> 44.2ms
 * 99 percentile frame time -> 100.0ms
 * error percentage -> ±1.5%
 * total hiccups (>0.05ms frames) -> 44 (6.72% of frames were hiccups)
 * total hiccups time (seconds) -> 3.076782
 * total frames -> 655
 * total frames time (seconds) -> 30.00663
```
### New

```
 * PERFORMANCE SCORE (0-100) -> 69
 * lowest frame time -> 35.3ms
 * average frame time -> 40.6ms
 * highest frame time -> 100.0ms
 * 1 percentile frame time -> 36.2ms
 * 50 percentile frame time -> 40.0ms
 * 99 percentile frame time -> 62.8ms
 * error percentage -> ±1.0%
 * total hiccups (>0.05ms frames) -> 14 (1.89% of frames were hiccups)
 * total hiccups time (seconds) -> 0.9778976
 * total frames -> 740
 * total frames time (seconds) -> 30.02143
```

## How to test

- Go to a scene with lots of UI elements like ICE Poker (-100,126) and make sure that they are correctly aligned when they get updated.
- Go to the same scene and before the UI loads, run outside the building and wait for all the tables to load, then enter the scene, the UI should appear without a massive hiccup

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

**Review by AjimenezDCL — APPROVED** [review]
Approving but left a suggestion, let me know what you think

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIShape.cs [hunk 1]
```diff
@@ -151,27 +154,65 @@ public class Model : BaseModel
...
 
         public UIShape parentUIComponent { get; protected set; }
-
-        private Coroutine layoutRefreshWatcher;
+        
+        private static Coroutine dirtyWatcher;
+        private static readonly Queue<UIShape> dirtyUIShapes = new Queue<UIShape>();
+        private static readonly List<UIShape> dirtyUIShapesNotReady = new List<UIShape>();
+        private const float DirtyWatcherUpdateBudget = 2/1000f;
+        private static readonly WaitForEndOfFrame eof = new WaitForEndOfFrame();
```
[the remark this answers is not in the document]
Comment (Kinerius) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIShape.cs:168]: Yes, this is a very good idea

File: unity-renderer/Assets/DCLPlugins/UIRefresherPlugin/UIRefresherController.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+using System.Collections.Generic;
+using DCL;
+using DCL.Components.Interfaces;
+using UnityEngine;
+
+namespace DCLPlugins.UIRefresherPlugin
+{
+    public class UIRefresherController : IDisposable
+    {
+        private const float DirtyWatcherUpdateBudget = 2/1000f;
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/DCLPlugins/UIRefresherPlugin/UIRefresherController.cs:11]:
```suggestion
        private const float DIRTY_WATCHER_UPDATE_BUDGET= 2/1000f;
```

File: unity-renderer/Assets/DCLPlugins/UIRefresherPlugin/UIRefresherController.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                while (queue.Count > 0 && CanRefreshMore(startTime))
+                {
+                    var uiShape = queue.Dequeue();
+                    uiShape.Refresh();
+                }
+            }
+        }
+        private static bool CanRefreshMore(float startTime)
+        {
+            if (Application.isBatchMode) return true;
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/DCLPlugins/UIRefresherPlugin/UIRefresherController.cs:70]: Great idea!! <3

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIShape.cs [hunk 4]
```diff
@@ -232,36 +232,34 @@ public string GetDebugName()
 
             return referencesContainer as T;
         }
-
-        IEnumerator LayoutRefreshWatcher()
+        
+        public virtual void RefreshAll()
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/UI/UIShape.cs:236]: Shouldnt be this named something like `RequestRefresh`?

**Review by AjimenezDCL — APPROVED** [review]
Awesome! left some small comments!

**Pull request merged**